Fix allocation bug in drizzle_binlog_get_filename

andreas-bok-sociomantic opened the pull request:
Allocate length of buffer to `|filename| + 1` to make
the string correctly '\0' terminated

Add test of `drizzle_binlog_get_filename` to the binlog unittest

andreas-bok-sociomantic committed:
Fix allocation bug in drizzle_binlog_get_filename

Allocate length of buffer to `|filename| + 1` to make
the string correctly '\0' terminated

ben-palmer-sociomantic commented on libdrizzle/binlog.cc:
```diff
@@ -493,9 +493,9 @@ void drizzle_binlog_get_filename(drizzle_st *con, char **filename, int file_inde
+    *filename = (char*)realloc(*filename, strlen(row[0]) + 1);
     memcpy(*filename, row[0], strlen(row[0]));
-    *filename[strlen(row[0])] = '\0';
+    (*filename)[strlen(row[0])] = '\0';
```
So previously this line would have been writing beyond the end of the filename array right?

andreas-bok-sociomantic replied:
Yeah, which should have segfaulted....

ben-palmer-sociomantic replied:
Is there a way to test the response is correct, perhaps in a unit test or something?

andreas-bok-sociomantic replied:
Yeah. I have changed the binlog unittest so it uses the filename retrieved from a `drizzle_binlog_get_filename` call

mathias-lang-sociomantic commented on libdrizzle/binlog.cc:
```diff
@@ -493,9 +493,9 @@ void drizzle_binlog_get_filename(drizzle_st *con, char **filename, int file_inde
+    *filename = (char*)realloc(*filename, strlen(row[0]) + 1);
     memcpy(*filename, row[0], strlen(row[0]));
-    *filename[strlen(row[0])] = '\0';
+    (*filename)[strlen(row[0])] = '\0';
```
`(*filename)[$-1] = '\0';` is way better

ben-palmer-sociomantic replied:
Can you do that in C?

mathias-lang-sociomantic replied:
I forgot about this inferior language, but @andreas-bok-sociomantic reminded me.
The main point was not to have double call to `strlen`, which is addressed now.

andreas-bok-sociomantic committed:
Add drizzle_binlog_get_filename to binlog unittest

Test the api function drizzle_binlog_get_filename
and use the retrieved filename when
starting the binlog

andreas-bok-sociomantic commented:
Updated based on PR comments

ben-palmer-sociomantic commented:
LGTM
